File: apps/android/app/src/main/java/com/signalasi/chat/CloudModelRequestRoutingPolicy.kt
package com.signalasi.chat

import org.json.JSONObject

internal object CloudModelRequestRoutingPolicy {
    const val DEEPSEEK_V4_PRO = "deepseek-v4-pro"
    const val DEEPSEEK_V4_FLASH = "deepseek-v4-flash"
    const val DEEPSEEK_V4_FLASH_VISION = "deepseek-v4-flash-vision-exp"

    private val deepSeekModels = listOf(
        AgentModelOption(DEEPSEEK_V4_PRO),
        AgentModelOption(DEEPSEEK_V4_FLASH),
        AgentModelOption(DEEPSEEK_V4_FLASH_VISION)
    )

    fun invocationProfile(contact: JSONObject): AgentInvocationProfile {
        val configured = buildList<AgentModelOption> {
            val models = contact.optJSONArray("cloud_models")
            if (models != null) {
                for (index in 0 until models.length()) {
                    val model = models.optJSONObject(index) ?: continue
                    val modelId = model.optString("model_id").trim()
                    if (modelId.isBlank() || any { it.id == modelId }) continue
                    add(AgentModelOption(
                        id = modelId,
                        displayName = modelId
                    ))
                }
            }
        }
        val models = ((if (isDeepSeek(contact)) deepSeekModels else emptyList()) + configured)
            .distinctBy(AgentModelOption::id)
        val current = contact.optString("selected_cloud_model")
            .ifBlank { contact.optString("cloud_model") }
        return AgentInvocationProfile(
            defaultModelId = current.takeIf { selected -> models.any { it.id == selected } }
                ?: models.firstOrNull()?.id.orEmpty(),
            models = models
        )
    }

    fun resolve(
        contact: JSONObject,
        requestedModelId: String,
        hasImageInput: Boolean
    ): JSONObject {
        val resolved = JSONObject(contact.toString())
        val profile = invocationProfile(resolved)
        val requested = requestedModelId.trim().takeIf { candidate ->
            profile.models.any { it.id == candidate }
        }
        val selectedModelId = requested
            ?: resolved.optString("selected_cloud_model").takeIf(String::isNotBlank)
            ?: resolved.optString("cloud_model")
        val effectiveModelId = effectiveModelId(selectedModelId, hasImageInput)
        if (!applyConfiguredModel(resolved, effectiveModelId)) {
            applyConfiguredModel(resolved, selectedModelId)
        }
        resolved.put("selected_cloud_model", selectedModelId)
        resolved.put("cloud_model", effectiveModelId)
        return resolved
    }

    fun effectiveModelId(selectedModelId: String, hasImageInput: Boolean): String =
        if (hasImageInput && selectedModelId in setOf(DEEPSEEK_V4_PRO, DEEPSEEK_V4_FLASH)) {
            DEEPSEEK_V4_FLASH_VISION
        } else {
            selectedModelId
        }

    private fun applyConfiguredModel(contact: JSONObject, modelId: String): Boolean {
        val models = contact.optJSONArray("cloud_models") ?: return false
        for (index in 0 until models.length()) {
            val model = models.optJSONObject(index) ?: continue
            if (model.optString("model_id") != modelId) continue
            contact.put("cloud_endpoint", model.optString("endpoint").ifBlank {
                contact.optString("cloud_endpoint")
            })
            contact.put("cloud_api_key", model.optString("api_key").ifBlank {
                contact.optString("cloud_api_key")
            })
            contact.put("cloud_api_style", model.optString("api_style").ifBlank {
                contact.optString("cloud_api_style").ifBlank { "openai" }
            })
            return true
        }
        return false
    }

    private fun isDeepSeek(contact: JSONObject): Boolean {
        val identity = buildString {
            append(contact.optString("cloud_provider"))
            append(' ')
            append(contact.optString("name"))
            append(' ')
            append(contact.optString("cloud_endpoint"))
            append(' ')
            append(contact.optString("cloud_model"))
        }.lowercase()
        return "deepseek" in identity
    }
}

File: apps/ios/SignalASI/CloudModelRequestRoutingPolicy.swift
import Foundation

enum CloudModelRequestRoutingPolicy {
  static let deepSeekV4Pro = "deepseek-v4-pro"
  static let deepSeekV4Flash = "deepseek-v4-flash"
  static let deepSeekV4FlashVision = "deepseek-v4-flash-vision-exp"

  static func invocationProfile(_ contact: SignalASIContact) -> AgentInvocationProfile {
    let options = models(for: contact).map {
      AgentModelOption(id: $0.modelId, displayName: $0.displayName.ifBlank($0.modelId))
    }
    let selected = contact.selectedCloudModelId.trimmingCharacters(in: .whitespacesAndNewlines)
    return AgentInvocationProfile(
      defaultModelId: options.contains(where: { $0.id == selected })
        ? selected
        : options.first?.id ?? "",
      models: options
    )
  }

  static func models(for contact: SignalASIContact) -> [CloudModelConfig] {
    guard isDeepSeek(contact), let template = contact.selectedCloudModel ?? contact.cloudModels.first else {
      return contact.cloudModels
    }
    let required = [deepSeekV4Pro, deepSeekV4Flash, deepSeekV4FlashVision]
    var models = required.map { modelId in
      contact.cloudModels.first(where: { $0.modelId == modelId })
        ?? derivedModel(modelId, template: template, contactId: contact.id)
    }
    for model in contact.cloudModels where !models.contains(where: { $0.modelId == model.modelId }) {
      models.append(model)
    }
    return models
  }

  static func model(in contact: SignalASIContact, modelId: String) -> CloudModelConfig? {
    let requested = modelId.trimmingCharacters(in: .whitespacesAndNewlines)
    if requested.isEmpty {
      return contact.selectedCloudModel
    }
    return models(for: contact).first { $0.modelId == requested }
  }

  static func resolve(
    contact: SignalASIContact,
    requestedModelId: String,
    hasImageInput: Bool
  ) -> SignalASIContact {
    let profile = invocationProfile(contact)
    let requested = requestedModelId.trimmingCharacters(in: .whitespacesAndNewlines)
    let selectedModelId = profile.models.contains(where: { $0.id == requested })
      ? requested
      : contact.selectedCloudModelId.ifBlank(profile.defaultModelId)
    let effectiveModelId = effectiveModelId(
      selectedModelId: selectedModelId,
      hasImageInput: hasImageInput
    )
    let availableModels = models(for: contact)
    guard let effective = availableModels.first(where: { $0.modelId == effectiveModelId })
      ?? availableModels.first(where: { $0.modelId == selectedModelId }) else {
      return contact
    }
    var resolved = contact
    resolved.cloudModels = availableModels
    resolved.selectedCloudModelId = effective.modelId
    return resolved
  }

  static func effectiveModelId(selectedModelId: String, hasImageInput: Bool) -> String {
    guard hasImageInput,
          [deepSeekV4Pro, deepSeekV4Flash].contains(selectedModelId) else {
      return selectedModelId
    }
    return deepSeekV4FlashVision
  }

  private static func isDeepSeek(_ contact: SignalASIContact) -> Bool {
    [
      contact.cloudProvider,
      contact.name,
      contact.displayName,
      contact.selectedCloudModel?.endpoint ?? "",
      contact.selectedCloudModelId
    ]
      .joined(separator: " ")
      .lowercased()
      .contains("deepseek")
  }

  private static func derivedModel(
    _ modelId: String,
    template: CloudModelConfig,
    contactId: String
  ) -> CloudModelConfig {
    let preset = CloudModelPreset.androidParity.first { $0.modelId == modelId }
    return CloudModelConfig(
      id: "\(contactId):\(modelId)",
      displayName: preset?.name ?? modelId,
      provider: template.provider.ifBlank("DeepSeek"),
      modelId: modelId,
      endpoint: preset?.endpoint ?? template.endpoint,
      apiStyle: preset?.apiStyle ?? template.apiStyle,
      keychainAccount: template.keychainAccount,
      updatedAt: template.updatedAt
    )
  }
}
